src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -597,6 +682,11 @@ impl_plan_tree_node_for_binary! { LogicalJoin }
 
 impl ColPrunable for LogicalJoin {
     fn prune_col(&self, required_cols: &[usize]) -> PlanRef {
+        // make `required_cols` point to internal table instead of output schema.
+        let required_cols = required_cols
+            .iter()
+            .map(|i| self.output_indices[*i])
+            .collect_vec();
```

Comment:
As discussed offline, let's try to make this bug fix a separate PR with test cases.